### PR TITLE
Update webpack 1 to 4 in billing-ui fix vulnerability in inovice-api and billing-invoice 

### PR DESCRIPTION
**Pull request by yuanhengzhao**
In general, I update the webpack version in billing-ui to 4; Fix the vulnerability in invoice-api and billing invoice; Fix a bug generated by map() object in python.

### **gitignore**
Update gitignore to avoid updating sensitive information and logs to git. Includes: logs/ config/s in each folder.

### **billing-api**
In _billing-api/billing_server/billing/__init__.py_

- line200: Map object in python can only be iterated once. To avoid it, add list() to map() object. 
- line 382: Convert it to string to avoid an error since type of retval.content is bytes and bytes doesn’t have find method.

In _billing-api/billing_server/billing/config/default.py_

- Move all the configuration in a separated file configs.py which is ignored by git to avoid potential information leaking.

In _billing-api/billing_server/billing/service/projects.py_

- line47: Add list() to map() object.

### **Billing-ui**

- Move babel.dev.js, babel.prod.js to webpack.config.dev.js and webpack.config.prod.js. They should be together for easier management.

In _webpack.config.dev.js and webpack.config.prod.js._

- Update webpack from 1 to 4 to fix several vulnerabilities.

In _billing-ui/scripts/build.js_

- line19: the stats get from webpack may also include some errors. Need to do errorHandling for them as well.
- line79: previous log information is not accurate.

In _billing-ui/src/index.js_

- line48: use proper expression for react component.

In 
_billing-ui/src/layouts/Base/Header/Header.scss_
_billing-ui/src/pages/Invoices/Invoices.js_
_billing-ui/src/pages/Invoices/Invoices.scss_
_billing-ui/src/pages/Login/Login.js_
_billing-ui/src/pages/Report/Report.js_
_billing-ui/src/pages/Report/Report.scss_
_billing-ui/src/routes.js_
_billing-ui/src/routes.js_
_billing-ui/src/services/email/sendEmail.js_
_billing-ui/src/services/invoices/fetchInvoices.js_
_billing-ui/src/services/projects/fetchProjects.js_
_billing-ui/src/services/reports/fetchReport.js_
_billing-ui/src/user.js_
_billing-ui/src/utils/fetchHeaders.js_ 

Change ~/** path to related path due to webpack updates.
### **Invoice-api**
In _invoice-api/src/server.ts._ 

- set configPath and authFilePath with default values so if they are not provided, avoid throwing error

### CODE REVIEW

File: billing-invoice/config.json.example [hunk 1]
```diff
@@ -18,6 +18,7 @@
     "rejectInsecure": true
   },
   "invoiceConfig": {
-    "api": "http://localhost:5000"
-  }
-}
+    "api": "http://localhost:4000"
+  },
+  "invoiceNumberPrefix": "OCG-"
+}
```
Comment (andricDu) [inline billing-invoice/config.json.example]: newline at end of file

File: billing-ui/src/services/reports/fetchReport.js [hunk 2]
```diff
@@ -37,3 +37,5 @@ export async function fetchReport ({projects, bucketSize, fromDate, toDate}) {
 
   return data;
 }
+
+export default fetchReport;
```
Comment (andricDu) [inline billing-ui/src/services/reports/fetchReport.js]: newline at EOF

File: billing-ui/src/services/projects/fetchProjects.js [hunk 3]
```diff
@@ -27,3 +27,4 @@ export async function fetchProjects () {
   if (response.status === 401) user.logout();
   return Promise.resolve(responseData);
 }
+export default fetchProjects;
```
Comment (andricDu) [inline billing-ui/src/services/projects/fetchProjects.js]: newline at EOF

File: billing-ui/src/utils/fetchHeaders.js [hunk 4]
```diff
@@ -14,11 +14,11 @@
...
+import { computed } from 'mobx';
+import user from '../user';
+const fetchHeaders = computed(() => ({
   Authorization: `Bearer ${user.token}`,
   Accept: 'application/json',
   'content-type': 'application/json',
-}));
\ No newline at end of file
+}));
+export default fetchHeaders;
```
Comment (andricDu) [inline billing-ui/src/utils/fetchHeaders.js]: newline at EOF

File: invoice-api/config.json.example [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+        "theme_font_name": "modern"
+      }
+    }
+  },
+  "emailRecipients": {
+    "oicr_finance_email": [],
+    "invoice_recipients": [],
+    "summary_recipients": []
+  }
+}
```
Comment (andricDu) [inline invoice-api/config.json.example]: newline at EOF

File: billing-ui/configs.js.example [hunk 6]
```diff
@@ -0,0 +1,5 @@
+var config = {
+    target: "http://localhost:5000"
+};
+
+module.exports = config;
```
Comment (andricDu) [inline billing-ui/configs.js.example]: newline at EOF

File: billing-api/billing_server/billing/config/default.py [hunk 7]
```diff
@@ -62,18 +48,4 @@
...
-    {
-        'period_start': '2017-06',
-        'period_end': '2017-06',
-        'discount': 0.7
-    }],
-    "oicr_demo_dusan" : [{
-        'discount': 0.8
-    }]
-}
+DISCOUNTS = configs.DISCOUNTS
```
Comment (caravinci) [inline billing-api/billing_server/billing/config/default.py]:
```suggestion
DISCOUNTS = configs.DISCOUNTS

```

File: billing-invoice/config.json.example (hunk 1 above)
Comment (caravinci) [inline billing-invoice/config.json.example]:
```suggestion
}

```

File: billing-ui/configs.js.example (hunk 6 above)
Comment (caravinci) [inline billing-ui/configs.js.example]:
```suggestion
module.exports = config;

```

File: billing-ui/src/services/projects/fetchProjects.js (hunk 3 above)
Comment (caravinci) [inline billing-ui/src/services/projects/fetchProjects.js]:
```suggestion
export default fetchProjects;

```

File: billing-ui/src/services/reports/fetchReport.js (hunk 2 above)
Comment (caravinci) [inline billing-ui/src/services/reports/fetchReport.js]:
```suggestion
export default fetchReport;

```

File: billing-ui/src/utils/fetchHeaders.js (hunk 4 above)
Comment (caravinci) [inline billing-ui/src/utils/fetchHeaders.js]:
```suggestion
export default fetchHeaders;

```

File: invoice-api/config.json.example (hunk 5 above)
Comment (caravinci) [inline invoice-api/config.json.example]:
```suggestion
}

```

File: billing-ui/src/routes.js [hunk 8]
```diff
@@ -27,19 +32,17 @@ function requireAuth(nextState, replace) {
...
       </Route>
-    </Router>
-)
+      <Route name="Test" path="test" component={Test} />
+    </Route>
+  </Router>)
 
-export default routes;
\ No newline at end of file
+export default Routes;
```
Comment (yuanhengzhao) [inline billing-ui/src/routes.js]:
```suggestion
export default Routes;

```

File: .gitignore [hunk 9]
```diff
@@ -203,4 +203,10 @@ invoice-api/node_modules/
...
 #log folder
-logs/
\ No newline at end of file
+logs/
+
+#Customize configuration
+configs.py
+configs.js
+billing-invoice/config.json
+invoice-api/config.json
```
Comment (yuanhengzhao) [inline .gitignore]:
```suggestion
invoice-api/config.json

```

**Pull request merged**